
<a id='ree'></a>
<div id="qe-notebook-header" style="text-align:right;">
        <a href="https://quantecon.org/" title="quantecon.org">
                <img style="width:250px;display:inline;" src="https://assets.quantecon.org/img/qe-menubar-logo.svg" alt="QuantEcon">
        </a>
</div>

# Equilíbrio das Expectativas Racionais

## Conteúdo

- [Equilíbrio das Expectativas Racionais](#Equilíbrio-das-Expectativas-Racionais)  
  - [Resumo](#Resumo)  
  - [Definindo o Equilíbrio das Expectativas Racionais](#Definindo-o-Equilíbrio-das-Expectativas-Racionais)  
  - [Computando um Equilíbrio](#Computando-um-Equilíbrio)  
  - [Exercícios](#Exercícios)  
  - [Soluções](#Soluções)  

> *Devidamente traduzido, revisado e adaptado do [QuantEcon](https://quantecon.org/) pelos bolsistas CNPq Pedro Luiz H. Furtado e Jonas Aragão M. Corpes, sob a supervisão do Prof. Christiano Penna, do CAEN/UFC.

> "Se você é tão inteligente, por que não é rico?"

## Resumo

Essa aula introduz o conceito de *equilíbrio das expectativas racionais*.

Para ilustrar, descrevemos uma versão quadrática linear de um modelo famoso e importante devido a Lucas e Prescott [[LP71]](https://julia.quantecon.org/../zreferences.html#lucasprescott1971).

Este artigo de 1971 é um de um pequeno número de artigos de pesquisa que iniciaram um *revolução nas expectativas racionais*.

Seguimos Lucas e Prescott empregando uma configuração que é prontamente "Bellmanizada" (ou seja, capaz de ser formulada em termos de problemas de programação dinâmica).

Como usamos configurações quadráticas lineares para demanda e custos, podemos adaptar as técnicas de programação LQ descritas [nessa aula](https://julia.quantecon.org/../dynamic_programming/lqcontrol.html).

Aprenderemos como o problema de um agente representativo difere do problema de um planejador e como um problema de planejamento pode ser usado para calcular quantidades de expectativas racionais.

Também aprenderemos sobre como um equilíbrio de expectativas racionais pode ser caracterizado como um [ponto fixo](https://en.wikipedia.org/wiki/Fixed_point_%28mathematics%29) de um mapeamento a partir de uma *lei do movimento percebido* para uma *lei do movimento real*.

A igualdade entre uma lei do movimento percebido e uma lei real para objetos endógenos, e que em todo o mercado captura em poucas palavras o que é o conceito de equilíbrio das expectativas racionais.

Finalmente, vamos aprender sobre o importante “truque grande $ K $, pequeno $ k $”, em um dispositivo de modelagem amplamente usado em macroeconomia.

Exceto que para nós:

- Em vez de “grande $ K $” será “grande $ Y $”  
- Em vez de “pequeno $ k $” será “pequeno $ y $”  

### O Truque: Grande $ Y $, Pequeno $ y $ 

Esse método amplamente usado se aplica em contextos em que uma “empresa representativa” ou agente é um “tomador de preço” operando dentro de um equilíbrio competitivo.

Queremos impor isso:

- A empresa representativa ou indivíduo toma o *agregado* $ Y $ como dado quando escolhe  $ y $, mas $ \ldots $.  
- No fim do dia, $ Y = y $, para que a firma representativa seja realmente representativa.


O truque do grande $ Y $, pequeno $ y $ realiza esses dois objetivos por:

- Tomando $ Y $ além do controle ao apresentar o problema de escolha de quem escolhe $ y $;  mas $ \ldots $.  
- Impondo $ Y = y $  *depois* de ter resolvido o problema de otimização individual.  


Por favor, observe como essa estratégia é aplicada à medida que a aula se desenrola.

Começamos aplicando o truque do grande $ Y $, pequeno $ y $ em um contexto estático muito simples.

#### Um exemplo estático muito simples do truque  Grande $ Y $, Pequeno $ y $ 

Considere um modelo estático no qual uma coleção de $ n $ produzem um bem homogêneo  que é vendido em um mercado competitivo.

Cada uma dessas $ n $ firmas vendem o produto $ y $.

O preço $ p $ do bem está em uma curva de demanda inversa


<a id='equation-ree-comp3d-static'></a>
$$
p = a_0 - a_1 Y \tag{1}
$$

onde

- $ a_i > 0 $ para $ i = 0, 1 $  
- $ Y = n y $ o nível de produção em todo o mercado


Cada firma tem a função de custo total:

$$
c(y) = c_1 y + 0.5 c_2 y^2,
\qquad c_i > 0 \text{ para } i = 1,2
$$

Os lucros de uma firma representativa são $ p y - c(y) $.

Usando [(1)](#equation-ree-comp3d-static), podemos expressar o problema da firma representativa como:


<a id='equation-max-problem-static'></a>
$$
\max_{y} \Bigl[ (a_0 - a_1 Y) y - c_1 y - 0.5 c_2 y^2 \Bigr] \tag{2}
$$

Ao colocar o problema [(2)](#equation-max-problem-static), queremos que a firma seja uma *tomadora de preços*.

Fazemos isso em relação a $ p $ e portanto $ Y $ é exógeno para a firma.

A essência do truque do grande $ Y $, pequeno $ y $ *não* é definir $ Y = n y $ *antes de* tomar a condição de primeira ordem em relação $ y $ no problema [(2)](#equation-max-problem-static).

Isso garante que a empresa seja uma tomadora de preço.

A condição de primeira ordem para o problema [(2)](#equation-max-problem-static) é:


<a id='equation-bigysimplefonc'></a>
$$
a_0 - a_1 Y - c_1 - c_2 y = 0 \tag{3}
$$

Neste ponto, *mas não antes*, substituimos $ Y = ny $ dentro [(3)](#equation-bigysimplefonc)
para obter a seguinte equação linear:


<a id='equation-staticy'></a>
$$
a_0 - c_1 - (a_1 + n^{-1} c_2) Y = 0 \tag{4}
$$

a ser resolvido para o produto do equilíbrio competitivo do mercado amplo $ Y $.

Depois de resolver para $ Y $, podemos computar o preço de equilíbrio competitivo $ p $ a partir de uma curva de demanda inversa [(1)](#equation-ree-comp3d-static).

### Leitura Adicional

As referências para essa aula incluem:

- [[LP71]](https://julia.quantecon.org/../zreferences.html#lucasprescott1971)  
- [[Sar87]](https://julia.quantecon.org/../zreferences.html#sargent1987), capítulo 14  
- [[LS18]](https://julia.quantecon.org/../zreferences.html#ljungqvist2012), capítulo 7  

### Configuração

In [1]:
using InstantiateFromURL
github_project("QuantEcon/quantecon-notebooks-julia", version = "0.5.0")
# github_project("QuantEcon/quantecon-notebooks-julia", version = "0.5.0", instantiate = true) # uncomment to force package installation

In [2]:
using LinearAlgebra, Statistics

## Definindo o Equilíbrio das Expectativas Racionais


<a id='index-1'></a>
Nossa primeira ilustração de um equilíbrio racional de expectativas envolve um mercado com $ n $ firmas, cada uma delas buscando maximizar o valor presente descontado dos lucros em face o ajustamento dos custos.

O ajustamento dos custos induzem as empresas a fazer ajustes graduais, o que, por sua vez, requer consideração de preços futuros.

As empresas individuais entendem que, através da curva de demanda inversa, o preço é determinado pelos valores fornecidos por outras empresas.

Portanto, cada empresa deseja prever os futuros suprimentos totais da indústria.

Em nosso contexto, uma previsão é gerada por uma crença sobre a lei do movimento para o estado agregado.

O equilíbrio das expectativas racionais prevalece quando essa crença coincide com a lei real do movimento gerada pelas escolhas de produção induzidas por essa crença.

Formulamos um equilíbrio racional de expectativas em termos de um ponto fixo de um operador que mapeia crenças em crenças ideais.

<a id='ree-ce'></a>

### Equilíbrio Competitivo com Ajustamento de Custos


<a id='index-2'></a>
Para ilustrar, considere uma coleção de $ n $ firmas produzindo um bem homogêneo que é vendido em um mercado competitivo.

Cada uma dessas $ n $ vendem o produto $ y_t $.

O preço $ p_t $ do bem está sobre uma curva de demanda inversa


<a id='equation-ree-comp3d'></a>
$$
p_t = a_0 - a_1 Y_t \tag{5}
$$

onde

- $ a_i > 0 $ para $ i = 0, 1 $  
- $ Y_t = n y_t $ é o nível de produção em todo o mercado.



<a id='ree-fp'></a>

#### O Problema da Firma

Cada firma é uma tomadora de preços.

Embora não exista incerteza, enfrentam os ajustamentos de custos.

Em particular, a firma escolhe um plano de produção para maximizar:

<a id='equation-ree-obj'></a>
$$
\sum_{t=0}^\infty \beta^t r_t \tag{6}
$$

onde


<a id='equation-ree-comp2'></a>
$$
r_t := p_t y_t - \frac{ \gamma (y_{t+1} - y_t )^2 }{2},
\qquad  y_0 \text{ dado} \tag{7}
$$

Em relação aos parâmetros,

- $ \beta \in (0,1) $ é o fator de desconto
- $ \gamma > 0 $ mede a taxa de ajustamento de custos do produto  


Em relação ao tempo, a firma observa $ p_t $ e $ y_t $ quando escolhe $ y_{t+1} $ no momento $ t $.

Para afirmar completamente o problema de otimização da empresa, exige que especifiquemos a dinâmica de todas as variáveis de estado.

Isso inclui aquelas com os quais a empresa se preocupa, mas não controla como  $ p_t $.

Agora nos voltamos para esse problema.

#### Preços e Produto Agregado

Em vista de [(5)](#equation-ree-comp3d), o incentivo da firma para prever o preço de mercado se traduz em um incentivo para prever a produção agregada $ Y_t $.

A produção agregada depende da escolha das outras firmas.

Assumimos que $ n $ é um número tão grande que a produção de uma única empresa tem um efeito insignificante na produção agregada.

Isso justifica que as empresas consideram que suas previsões de produção agregada não são afetadas por suas próprias decisões de produção.

#### A Crença da Firma

Supomos que a firma acredita que a produção em todo o mercado $ Y_t $ segue a lei de movimento


<a id='equation-ree-hlom'></a>
$$
Y_{t+1} =  H(Y_t) \tag{8}
$$

onde $ Y_0 $ é uma condição inicial conhecida.

A *função de crença* $ H $ é um objeto de equilíbrio e, portanto, ainda precisa ser determinado.

#### Comportamento Ótimo diante das Crenças

Agora vamos considerar uma crença específica $ H $ em [(8)](#equation-ree-hlom) e investigar a resposta da firma a isso.

Se $ v $ é a função de valor ótimo para o problema da firma dado $ H $.

A função de valor satisfaz a equação de Bellman


<a id='equation-comp4'></a>
$$
v(y,Y) = \max_{y'} \left\{ a_0 y - a_1 y Y - \frac{ \gamma (y' - y)^2}{2}   + \beta v(y', H(Y))\right\} \tag{9}
$$

Vamos denotar a função política ótima da firma por $ h $, de modo a:


<a id='equation-comp9'></a>
$$
y_{t+1} = h(y_t, Y_t) \tag{10}
$$

onde


<a id='equation-ree-opbe'></a>
$$
h(y, Y) := \mathop{\mathrm{arg\,max}}_{y'}
\left\{ a_0 y - a_1 y Y - \frac{ \gamma (y' - y)^2}{2}   + \beta v(y', H(Y))\right\} \tag{11}
$$

Evidentemente $ v $ e $ h $ ambos dependem de $ H $.

#### Caracterização de Primeira Ordem de $ h $

A seguir, será útil ter uma segunda caracterização de $ h $, baseada nas condições de primeira ordem.

A condição de primeira ordem necessária para escolher $ y' $ é:


<a id='equation-comp5'></a>
$$
-\gamma (y' - y) + \beta v_y(y',H(Y)) =0 \tag{12}
$$

Um resultado importante usual do envelope de Benveniste-Scheinkman  [[BS79]](https://julia.quantecon.org/../zreferences.html#benvenistescheinkman1979) implica que
diferenciar $ v $ com respeito a $ y $ podemos ingenuamente diferenciar o lado direito de [(9)](#equation-comp4), dado:

$$
v_y(y,Y) = a_0 - a_1 Y + \gamma (y' - y)
$$

Substituindo essa equação em [(12)](#equation-comp5) dá a *equação de Euler*


<a id='equation-ree-comp7'></a>
$$
-\gamma (y_{t+1} - y_t) + \beta [a_0 - a_1 Y_{t+1} + \gamma (y_{t+2} - y_{t+1} )] =0 \tag{13}
$$

A firma otimiza um caminho de produção que satisfaça[(13)](#equation-ree-comp7), tomando [(8)](#equation-ree-hlom) como dado, e sujeito a:

- as condições iniciais para $ (y_0, Y_0) $  
- a condição terminal $ \lim_{t \rightarrow \infty } \beta^t y_t v_y(y_{t}, Y_t) = 0 $  


A última condição é chamada de *condição de transversalidade*, e atua como uma condição de primeira ordem necessária  “no infinito”.

A regra de decisão da firma resolve a equação da diferença [(13)](#equation-ree-comp7) sujeito à condição incial dada $ y_0 $ e a condição de transversalidade.

Observe que resolvendo a equação de Bellman [(9)](#equation-comp4) para $ v $ e então $ h $ em [(11)](#equation-ree-opbe)produz uma regra de decisão que impõe automaticamente a equação de Euler [(13)](#equation-ree-comp7) e a condição de transversalidade.

#### A Lei do Movimento Atual para $ \{Y_t\} $

Como visto, uma determinada crença se traduz em uma regra de decisão específica $ h $.

Lembrando que $ Y_t = ny_t $, a *lei de movimento atual* para a produção em todo o mercado é então:


<a id='equation-ree-comp9a'></a>
$$
Y_{t+1} = n h(Y_t/n, Y_t) \tag{14}
$$

Então, quando a firma acredita que a lei de movimento para o produto de todo o mercado é [(8)](#equation-ree-hlom), seu comportamento otimizador torna a lei do movimento real [(14)](#equation-ree-comp9a).


<a id='ree-def'></a>

### Definição doEquilíbrio das Expectativas Racionais

Um *equilíbrio de expectativas racionais* ou *equilíbrio competitivo recursivo* do modelo com ajsutamento de custos é uma regra de decisão $ h $ e uma lei de movimento agregado $ H $ de modo que:

1. Dado a crença $ H $, o mapa $ h $ é a função de política ótima da firma.  
1. A lei de movimento $ H $ satisfaz $ H(Y)= nh(Y/n,Y) $ para todo.
  $ Y $  


Então, um equilíbrio das expectativas racionais iguala as leis de movimento percebidas e atuais [(8)](#equation-ree-hlom) e [(14)](#equation-ree-comp9a).

#### Caracterização do Ponto Fixo

Como vimos, o problema ótimo da empresa induz um mapeamento $ \Phi $ a partir de uma lei de movimento percebida $ H $ para a produção em todo o mercado de acordo com uma lei de movimento atual $ \Phi(H) $.

O mapeamento $ \Phi $ é um composição de duas operações, levando uma lei de movimento percebida para uma regra de decisão via [(9)](#equation-comp4)–[(11)](#equation-ree-opbe), e uma regra de decisão para a lei atual via [(14)](#equation-ree-comp9a).

O componente $ H $ de um equilíbrio das expectativas racionais é um ponto fixo de $ \Phi $.

## Computando um Equilíbrio


<a id='index-3'></a>
Agora vamos considerar o problema do cálculo do equilibrio das expectativas racionais.

### Comportamento Inadequado de $ \Phi $

Os leitores acostumados a argumentos de programação dinâmica podem tentar resolver esse problema escolhendo alguma hipótese $ H_0 $ para agregar a lei do movimento e então iterar com $ \Phi $.

Infelizmente, o mapeamento  $ \Phi $ não é uma contração.

Em particular, não há garantias que iterações diretas em $ \Phi $ convergem <sup><a href=#fn-im id=fn-im-link>[1]</a></sup>.

Felizmente,  existe outro método que funciona aqui.

O método explora uma conexão geral entre equilíbrio e ótimo de Pareto expressado nos teoremas fundamentais da economia do bem-estar (veja, por exemplo, [[MCWG95]](https://julia.quantecon.org/../zreferences.html#mcwg1995)).

Lucas e Prescott [[LP71]](https://julia.quantecon.org/../zreferences.html#lucasprescott1971)usaram esse método para construir um equilíbrio das expectativas racionais.

Os detalhes a seguir.


<a id='ree-pp'></a>

### Uma Abordagem do Problema do Planejador


<a id='index-4'></a>
Nosso plano de ataque é combinar as equações de Euler do problema de mercado com as de um problema de escolha de agente único.

Como veremos, esse problema de planejadot pode ser resolvido pelo controle LQ ([o regulador linear](https://julia.quantecon.org/../dynamic_programming/lqcontrol.html)).

As quantidades ótimas do problema de planejador são as quantidades do equilíbrio das expectativas racionais.

O preço de equilíbrio das expectativas racionais pode ser obtido como um preço mínimo no problema de planejador.

Por conveniência, nesta seção, definimos $ n=1 $.

Primeiro calculamos uma soma do excedente do consumidor e do produtor no momento $ t $


<a id='equation-comp10'></a>
$$
s(Y_t, Y_{t+1})
:= \int_0^{Y_t} (a_0 - a_1 x) \, dx - \frac{ \gamma (Y_{t+1} - Y_t)^2}{2} \tag{15}
$$

O primeiro termo é a área sob a curva de demanda, enquanto o segundo mede os custos sociais da mudança de produção.

O *problema do planejador* é escolher um plano de produção $ \{Y_t\} $ para maximizar:

$$
\sum_{t=0}^\infty \beta^t s(Y_t, Y_{t+1})
$$

sujeito a uma condição inicial para $ Y_0 $.

### Solução para o Problema do Planejador

Avaliando a integral em [(15)](#equation-comp10) produz a forma quadrática $ a_0
Y_t - a_1 Y_t^2 / 2 $.

Como um resultado, a equação de Bellman para um problema de planejador é:


<a id='equation-comp12'></a>
$$
V(Y) = \max_{Y'}
\left\{a_0  Y - {a_1 \over 2} Y^2 - \frac{ \gamma (Y' - Y)^2}{2} + \beta V(Y') \right\} \tag{16}
$$

A condição de primeira ordem associada é:


<a id='equation-comp14'></a>
$$
-\gamma (Y' - Y) + \beta V'(Y') = 0 \tag{17}
$$

Aplicando a mesma fórmula de Benveniste-Scheinkman temos:

$$
V'(Y) = a_0 - a_1 Y + \gamma (Y' - Y)
$$

Substituindo isso na equação [(17)](#equation-comp14) e reorganizando leva à equação de Euler


<a id='equation-comp16'></a>
$$
\beta a_0 + \gamma Y_t - [\beta a_1 + \gamma (1+ \beta)]Y_{t+1} + \gamma \beta Y_{t+2} =0 \tag{18}
$$

### O *Insight* principal

Retorne a equação [(13)](#equation-ree-comp7) e defina $ y_t = Y_t $ para todo $ t $.

(lembre-se que para essa seção definimos $ n=1 $ para simplificar os cálculos)

Uma pequena quantidade de álgebra o convencerá de que quando $ y_t=Y_t $, as equações [(18)](#equation-comp16) e [(13)](#equation-ree-comp7) são idênticas.

Assim, a equação de Euler para o problema de planejamento corresponde à equação da diferença de segunda ordem que derivamos por:

1. Encontrando a equação de Euler da firma representativa, e 
2. substituindo nela a expressão  $ Y_t = n y_t $ que “faz a firma representativa ser representativa”.  


Se for apropriado aplicar as mesmas condições terminais para essas duas equações diferenciais, que é, então verificamos que uma solução do problema de planejamento também é uma sequência de quantidades do equilíbrio das expectativas racionais.

Segue-se que, para este exemplo, podemos calcular quantidades de equilíbrio formando o problema ótimo do regulador linear correspondente à equação de Bellman [(16)](#equation-comp12).

A função de política ótima para o problema de planejamento é a lei de movimento agregada
$ H $ que a empresa representativa enfrenta dentro de um equilíbrio das expectativas racionais.

#### Estrutura da Lei de Movimento

Como você é solicitado a mostrar nos exercícios, o fato de o problema do planejador ser um problema de LQ implica uma política ideal - e, portanto, lei agregada do movimento - assume a forma:


<a id='equation-ree-hlom2'></a>
$$
Y_{t+1}
= \kappa_0 + \kappa_1 Y_t \tag{19}
$$

para algum par de parâmetros $ \kappa_0, \kappa_1 $.

Agora que sabemos que a lei agregada do movimento é linear, podemos ver pela equação de Bellman da firma [(9)](#equation-comp4) que o problema da firma também pode ser enquadrado como um problema de LQ.

Como você é solicitado a mostrar nos exercícios, a formulação LQ do problema da empresa implica uma lei do movimento que se parece com a seguinte:


<a id='equation-ree-ex5'></a>
$$
y_{t+1} = h_0 + h_1 y_t + h_2 Y_t \tag{20}
$$

Portando o equilíbrio das expectativas racionais será definido pelos parâmetros
$ (\kappa_0, \kappa_1, h_0, h_1, h_2) $ em [(19)](#equation-ree-hlom2)–[(20)](#equation-ree-ex5).

## Exercícios


<a id='ree-ex1'></a>

### Exercício 1

Considere o problema da firma [descrito acima](#ree-fp).

Seja a função de crença da firma $ H $ como dada em [(19)](#equation-ree-hlom2).

Formule o problema da firma como um problema regulador linear ótimo com desconto, tomando cuidado para descrever todos os objetos necessários.

Use o tipo `LQ` disponível no pacote [QuantEcon.jl](http://quantecon.org/quantecon-jl) para resolver o problema da empresa para os seguintes valores de parâmetro:

$$
a_0= 100, a_1= 0.05, \beta = 0.95, \gamma=10, \kappa_0 = 95.5, \kappa_1 = 0.95
$$

Expresse a solução do problema da empresa na forma [(20)](#equation-ree-ex5) e forneça os valores para cada $ h_j $.

Se houvesse $ n $ empresas concorrentes idênticas, todas se comportando de acordo com [(20)](#equation-ree-ex5), o que [(20)](#equation-ree-ex5) implicaria a lei de movimento *real* [(8)](#equation-ree-hlom) para a oferta de mercado.


<a id='ree-ex2'></a>

### Exercício 2 

Considere os seguintes pares $ \kappa_0, \kappa_1 $ como candidatos ao componente da lei do movimento agregado de um equilíbrio das expectativas racionais (veja
[(19)](#equation-ree-hlom2)).

Estendendo o programa que você escreveu para o exercício 1, determine quais deles satisfazem [a definição](#ree-def) de um equilíbrio das expectativas racionais

- (94.0886298678, 0.923409232937)  
- (93.2119845412, 0.984323478873)  
- (95.0818452486, 0.952459076301)  


Descreva um algoritmo iterativo que usa o programa que você escreveu para o exercício 1 para calcular um equilíbrio das expectativas racionais.

(Na verdade, você não está sendo solicitado a usar o algoritmo que está sugerindo)


<a id='ree-ex3'></a>

### Exercício 3

Lembre-se do problema do planejador [descrito acima](#ree-pp)

1. Formule o problema do planejador como um problema de LQ.
1. Resolva-o usando os mesmos valores de parâmetro no exercício 1: 
  
  - $ a_0= 100, a_1= 0.05, \beta = 0.95, \gamma=10 $  
  
1. Represente a solução na forma $ Y_{t+1} = \kappa_0 + \kappa_1 Y_t $.  
1. Compare suas respostas com os resultado do exercício 2.  



<a id='ree-ex4'></a>

### Exercício 4

Um monopolista enfrenta a curva de demanda da indústria [(5)](#equation-ree-comp3d) e escolhe $ \{Y_t\} $ para maximizar $ \sum_{t=0}^{\infty} \beta^t r_t $ onde

$$
r_t = p_t Y_t - \frac{\gamma (Y_{t+1} - Y_t)^2 }{2}
$$

Formule esse problema como um problema de LQ.

Calcule a política ideal usando os mesmos parâmetros do exercício anterior.

Em particular, resolva os parâmetros em:

$$
Y_{t+1} = m_0 + m_1 Y_t
$$

Compare seus resultados com o exercício anterior. Comente.

## Soluções

### Exercício 1

In [3]:
using QuantEcon, Printf, LinearAlgebra

Para mapear um problema em um [problema de controle linear com desconto](http://quant-econ.net/jl/lqcontrol.html), precisamos definir:

- o vetor de estado $ x_t $ e o vetor de controle $ u_t $  
- as matrizes $ A, B, Q, R $ que definem as preferências e a lei de movimento para o estado 


Para os vetores de estado e controle escolhemos:

$$
x_t = \begin{bmatrix} y_t \\ Y_t \\ 1 \end{bmatrix},
\qquad
u_t = y_{t+1} - y_{t}
$$

Para $ , B, Q, R $ definimos:

$$
A =
\begin{bmatrix}
    1 & 0 & 0 \\
    0 & \kappa_1 & \kappa_0 \\
    0 & 0 & 1
\end{bmatrix},
\quad
B = \begin{bmatrix} 1 \\ 0 \\ 0 \end{bmatrix} ,
\quad
R =
\begin{bmatrix}
    0 & a_1/2 & -a_0/2 \\
    a_1/2 & 0 & 0 \\
    -a_0/2 & 0 & 0
\end{bmatrix},
\quad
Q = \gamma / 2
$$

Ao multiplicar, você pode confirmar que:

- $ x_t' R x_t + u_t' Q u_t = - r_t $  
- $ x_{t+1} = A x_t + B u_t $  


Usaremos o módulo  `lqcontrol.jl` para resolver o problema da firma nos valores de parâmetros para o estado.

Isso retornará uma política LQ $ F $ com a interpretação:
$ u_t = - F x_t $, ou

$$
y_{t+1} - y_t = - F_0 y_t - F_1 Y_t - F_2
$$

Correspondência de parâmetros com  $ y_{t+1} = h_0 + h_1 y_t + h_2 Y_t $ leva a:

$$
h_0 = -F_2, \quad h_1 = 1 - F_0, \quad h_2 = -F_1
$$

Aqui está a nossa solução.

In [4]:
# modelo parâmetros
a0 = 100
a1 = 0.05
β = 0.95
γ = 10.0

# crenças
κ0 = 95.5
κ1 = 0.95

# formulando o problema LQ
A = [1  0  0
     0 κ1 κ0
     0  0  1]

B = [1.0, 0.0, 0.0]

R = [    0   a1/2 -a0/2
      a1/2      0     0
     -a0/2      0     0]

Q = 0.5 * γ

# resolver para a política ótima
lq = QuantEcon.LQ(Q, R, A, B; bet = β)
P, F, d = stationary_values(lq)

hh = h0, h1, h2 = -F[3], 1 - F[1], -F[2]

@printf("F = [%.3f, %.3f, %.3f]\n", F[1], F[2], F[3])
@printf("(h0, h1, h2) = [%.3f, %.3f, %.3f]\n", h0, h1, h2)

F = [-0.000, 0.046, -96.949]
(h0, h1, h2) = [96.949, 1.000, -0.046]


A implicação é que

$$
y_{t+1} = 96.949 + y_t - 0.046 \, Y_t
$$

Para o caso $ n > 1 $, lembre-se que $ Y_t = n y_t $, que,
combinado com a equação anterior, produz:
$$
Y_{t+1}
= n \left( 96.949 + y_t - 0.046 \, Y_t \right)
= n 96.949 + (1 - n 0.046) Y_t
$$

### Exercício 2

Para determinar se um par $ \kappa_0, \kappa_1 $ forma a componente lei do movimento agregado de um equilíbrio das expectativas racionais, podemos proceder da seguinte forma:

- Determine a lei de movimento correspondente:
  $ y_{t+1} = h_0 + h_1 y_t + h_2 Y_t $.  
- Teste se a lei agregada associada:
  :$ Y_{t+1} = n h(Y_t/n, Y_t) $ avalia para
  $ Y_{t+1} = \kappa_0 + \kappa_1 Y_t $.  


No segundo passo podemos usar $ Y_t = n y_t = y_t $, para que
$ Y_{t+1} = n h(Y_t/n, Y_t) $ se torne:

$$
Y_{t+1} = h(Y_t, Y_t) = h_0 + (h_1 + h_2) Y_t
$$

Portanto para testar o segundo passo podemos testar $ \kappa_0 = h_0 $ e
$ \kappa_1 = h_1 + h_2 $.

O código seguinte implementa esse teste.

In [5]:
candidates = ([94.0886298678, 0.923409232937],
              [93.2119845412, 0.984323478873],
              [95.0818452486, 0.952459076301])

for (k0, k1) in candidates
    A = [1  0  0
         0 k1 k0
         0  0  1]
    lq = QuantEcon.LQ(Q, R, A, B; bet=β)
    P, F, d = stationary_values(lq)
    hh = h0, h1, h2 = -F[3], 1 - F[1], -F[2]

    if isapprox(k0, h0; atol = 1e-4) && isapprox(k1, h1 + h2; atol = 1e-4)
        @printf("Equilibrium pair = (%.6f, %.6f)\n", k0, k1)
        @printf("(h0, h1, h2) = [%.6f, %.6f, %.6f]\n", h0, h1, h2)
    end
end

Equilibrium pair = (95.081845, 0.952459)
(h0, h1, h2) = [95.081891, 1.000000, -0.047541]


A resultado nos diz que a resposta é o par (iii), que implica
$ (h_0, h_1, h_2) = (95.0819, 1.0000, -.0475) $.

(Observe que usamos `isapprox` para testar a igualdade de números de ponto flutuante, pois a igualdade exata é muito rigorosa)

Em relação ao algoritmo iterativo, pode-se fazer um loop de um dado par
$ (\kappa_0, \kappa_1) $ para a lei da firma associada e então para um novo $ (\kappa_0, \kappa_1) $.

Isso equivale à implementação do operador  $ \Phi $ descrito na aula.

(Em geral, não há garantia de que esse processo iterativo convirja para um equilíbrio racional de expectativas)

### Exercício 3

Somos solicitados a escrever o problema do planejador como um problema de LQ.

Para os vetores de estado e controle escolhemos:

$$
x_t = \begin{bmatrix} Y_t \\ 1 \end{bmatrix},
\quad
u_t = Y_{t+1} - Y_{t}
$$

Para as matrizes LQ definimos:

$$
A = \begin{bmatrix} 1 & 0 \\ 0 & 1 \end{bmatrix},
\quad
B = \begin{bmatrix} 1 \\ 0 \end{bmatrix},
\quad
R = \begin{bmatrix} a_1/2 & -a_0/2 \\ -a_0/2 & 0 \end{bmatrix},
\quad
Q = \gamma / 2
$$

Ao multiplicar, você pode confirmar que

- $ x_t' R x_t + u_t' Q u_t = - s(Y_t, Y_{t+1}) $  
- $ x_{t+1} = A x_t + B u_t $  


Obtendo a política ótima e usando $ u_t = - F x_t $ ou

$$
Y_{t+1} - Y_t = -F_0 Y_t - F_1
$$

podemos obter a lei agregada implícita do movimento via
$ \kappa_0 = -F_1 $ and $ \kappa_1 = 1-F_0 $

In [6]:
# Formule o problema do planejador LQ
A = I + zeros(2, 2)
B = [1.0, 0.0]

R = [ a1 / 2.0  -a0 / 2.0
     -a0 / 2.0        0.0]

Q = γ / 2.0

# resolva para a política ótima
lq = QuantEcon.LQ(Q, R, A, B; bet=β)
P, F, d = stationary_values(lq)

# imprima os resultados
κ0, κ1 = -F[2], 1 - F[1]
println("κ0=$κ0\tκ1=$κ1")

κ0=95.08187459215002	κ1=0.9524590627039248


O resultado produz o mesmo $ (\kappa_0, \kappa_1) $ par obtido como um equilíbrio do exercício anterior.

### Exercício 4

O problema do monopolista LQ é quase idêntico ao problema do planejador
do exercício anterior, exceto que

$$
R = \begin{bmatrix}
    a_1 & -a_0/2 \\
    -a_0/2 & 0
\end{bmatrix}
$$

O problema pode ser resolvido como o seguinte

In [7]:
# formule o problema do monopolista LQ
A = I + zeros(2, 2)
B = [1.0, 0.0]

R = [       a1   -a0 / 2.0
     -a0 / 2.0         0.0]

Q = γ / 2.0

# Resolva para a política ótima
lq = QuantEcon.LQ(Q, R, A, B; bet=β)
P, F, d = stationary_values(lq)

# Imprima os resultados
m0, m1 = -F[2], 1 - F[1]
println("m0=$m0\tm1=$m1")

m0=73.47294403502833	m1=0.9265270559649701


Vemos que a lei de movimento para o monopolista é aproximadamente 
$ Y_{t+1} = 73.4729 + 0.9265 Y_t $.

No caso das expectativas racionais, a lei do movimento era aproximadamente 
$ Y_{t+1} = 95.0819 + 0.9525 Y_t $.

Uma maneira de comparar essas duas leis do movimento é por seus pontos fixos, que fornecem um produto de equilíbrio a longo prazo em cada caso.

Para leis de forma $ Y_{t+1} = c_0 + c_1 Y_t $, o ponto fixo é:
$ c_0 / (1 - c_1) $.

Se você analisar os números, verá que o monopolista adota uma quantidade menor a longo prazo do que a obtida pelo mercado competitivo, implicando um preço de mercado mais alto.

Isso é análogo aos resultados elementares de casos estáticos.

**Notas de Rodapé**

<p><a id=fn-im href=#fn-im-link><strong>[1]</strong></a> Um literatura que estuda se os modelos povoadas com agentes que aprendem, podem convergir aos equilíbrios das expectativas racionais apresenta iterações em uma modificação do mapeamento $ \Phi $ que pode ser aproximada como $ \gamma \Phi + (1-\gamma)I $. Aqui $ I $ é o operador identidade e $ \gamma \in (0,1) $ é um *parâmetro de relaxamento*.
Veja [MS89](https://julia.quantecon.org/../zreferences.html#marcetsargent1989) e [[EH01]](https://julia.quantecon.org/../zreferences.html#evanshonkapohja2001) para declarações e aplicações desta abordagem para estabelecer condições sob as quais coleções de agentes adaptativos que usam o aprendizado de mínimos quadrados convergem para um equilíbrio das expectatívas racionais.